Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs

```diff
@@ -39,43 +33,50 @@ public BlockStoreInnerStepFindBlocks(ILoggerFactory loggerFactory)
         public override async Task<InnerStepResult> ExecuteAsync(BlockStoreInnerStepContext context)
         {
             this.logger.LogTrace("()");
-            context.GetNextBlock();
 
-            if (await ShouldStopFindingBlocks(context))
+            while (await ShouldStopFindingBlocks(context) == false)
             {
-                if (!context.DownloadStack.Any())
-                {
-                    this.logger.LogTrace("(-):{0}", InnerStepResult.Stop);
-                    return InnerStepResult.Stop;
-                }
-
-                context.StopFindingBlocks();
-            }
-            else
-            {
-                context.BlockStoreLoop.BlockPuller.AskBlock(context.NextChainedBlock);
                 context.DownloadStack.Enqueue(context.NextChainedBlock);
-
-                if (context.DownloadStack.Count == context.BlockStoreLoop.BatchDownloadSize)
-                    context.StopFindingBlocks();
+                context.GetNextBlock();
             }
 
+            context.StopFindingBlocks();
+
+            if (context.DownloadStack.Any())
+                context.BlockStoreLoop.BlockPuller.AskForMultipleBlocks(context.DownloadStack.ToArray());
+
             this.logger.LogTrace("(-):{0}", InnerStepResult.Next);
+
             return InnerStepResult.Next;
         }
 
         private async Task<bool> ShouldStopFindingBlocks(BlockStoreInnerStepContext context)
         {
             this.logger.LogTrace("()");
 
-            bool res = (context.NextChainedBlock == null)
-                || (context.NextChainedBlock.Header.HashPrevBlock != context.InputChainedBlock.HashBlock)
-                || (context.NextChainedBlock.Height > context.BlockStoreLoop.ChainState.HighestValidatedPoW?.Height)
-                || (context.BlockStoreLoop.PendingStorage.ContainsKey(context.NextChainedBlock.HashBlock))
-                || (await context.BlockStoreLoop.BlockRepository.ExistAsync(context.NextChainedBlock.HashBlock));
+            bool result = false;
+
+            if (context.NextChainedBlock == null)
+                result = true;
```

Comment:
Well, I'm not really happy about your decision here, in fact I think the code that you just refactored to is actually worse than what you started with (that one with single return and logging). It seems the very nature of the whole mechanism will have a log gap and I don't understand why (to save lines of code or to have a bit easier debugging?). I'd be absolutely happy with that separated logging with multiple returns giving the context of why it returned true. But we seem to disagree there.

I'll leave it as you think it is better. So I will just run the tests to check the functionality is not somehow broken and approve it.